pkg/audit/service.go

```diff
@@ -34,25 +55,52 @@ func NewService(ctx context.Context, statDBPort string, maxRetries int, pointers
 	if err != nil {
 		return nil, err
 	}
-	return &Service{Cursor: cursor, Verifier: verifier, Reporter: reporter}, nil
+	localCtx, localCancel := context.WithCancel(ctx)
+	return &Service{Cursor: cursor,
+		Verifier: verifier,
+		Reporter: reporter,
+		ctx:      localCtx,
+		cancel:   localCancel,
+		errs:     []error{},
+	}, nil
 }
 
 // Run calls Cursor and Verifier to continuously request random pointers, then verify data correctness at
 // a random stripe within a segment
-func (service *Service) Run(ctx context.Context) (err error) {
-	// TODO(James): make this function run indefinitely instead of once
-	stripe, err := service.Cursor.NextStripe(ctx)
-	if err != nil {
-		return err
-	}
-	verifiedNodes, err := service.Verifier.verify(ctx, stripe.Index, stripe.Segment)
-	if err != nil {
-		return err
-	}
-	err = service.Reporter.RecordAudits(ctx, verifiedNodes)
-	// TODO: if Error.Has(err) then log the error because it means not all node stats updated
-	if !Error.Has(err) && err != nil {
-		return err
-	}
-	return nil
+func (service *Service) Run(ctx context.Context, interval time.Duration) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	zap.S().Info("Audit cron is starting up")
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				stripe, err := service.Cursor.NextStripe(ctx)
+				if err != nil {
+					service.errs = append(service.errs, err)
+					service.cancel()
+				}
+				verifiedNodes, err := service.Verifier.verify(ctx, stripe.Index, stripe.Segment)
+				if err != nil {
+					service.errs = append(service.errs, err)
+					service.cancel()
+				}
+				err = service.Reporter.RecordAudits(ctx, verifiedNodes)
+				// TODO: if Error.Has(err) then log the error because it means not all node stats updated
+				if err != nil {
+					service.errs = append(service.errs, err)
+					service.cancel()
+				}
+			case <-service.ctx.Done():
+				return
+			}
+		}
+	}()
+	return utils.CombineErrors(service.errs...)
```

Comment:
so this function doesn't block does it?

Reply:
James says he assumes so but we don't know 😬

Reply:
Will look to add `collectErrors()`